go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,259 @@
+package libovsdbops
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+)
+
+const (
+	// db ownership ExternalIDs, shared by all controllers
+	// OwnerController and OwnerObjectType define managing entity (a.k.a. owner) for given db object
+	OwnerController = types.OvnK8sPrefix + "/controller"
+	OwnerObjectType = types.OvnK8sPrefix + "/object-type"
+	// SecondaryIndexKey is a part of a secondary index, together with OwnerController and OwnerObjectType
+	// May be used by controllers to store e.g. namespace+name of the object.
+	SecondaryIndexKey = types.OvnK8sPrefix + "/name"
+	// PrimaryIDKey will be used as a primary index, that is unique for every db object,
+	// and can be built based on the combination of all the other ids.
+	PrimaryIDKey = types.OvnK8sPrefix + "/id"
+
+	// Owner Object Types
+	EgressFirewallDNSOwnerType = "EgressFirewallDNS"
+	EgressQoSOwnerType         = "EgressQoS"
+	NetworkPolicyOwnerType     = "NetworkPolicy"
+	NamespaceOwnerType         = "Namespace"
+	HybridNodeRouteOwnerType   = "HybridNodeRoute"
+)
+
+// DbIndex is a structure representing a set of db object ExternalIDs, used to uniquely identify
```

Comment:
I was hoping this would become simpler, and not more complicated like its turning out to be.

This would be more similar to my idea:
```

type DBObjectType string
const AddressSetType DBObjectType = "AddressSetType"

// by themselves, these are also secondary ids
var AddressSetTypePrimaryIDs = []string{"controller", "name", "family"}
// other secondary ids that are not part of the primary ids
var AddressSetTypeOtherIDs = nil

// I like this new name you came up with (more than DBIndex)
// This could be a struct, depending on what other information 
// we want to have (i.e. lazy initialization and caching of unique id, 
// quick tracking of object type, etc...)
type DBObjectIDs map[string]string

func NewDBObjectIDsFromExternalIDs(externalIDs map[string]string) DBObjectIDs {
// one of the provided external ids should identify the object type
}

func NewDBObjectIDsForObjectType(type DBObjectType, ids map[string]string) DBObjectIDs {
// select ids for that object type
// we can have slight variations of this method if we want to verify all, 
// primary or secondary ids are provided
}

func (id DBObjectIDs) GetUniqueID() string {
// ...
}

func (id DBObjectIDs) GetExternalIDs() map[string]string {
// ...
}

func (id DBObjectIDs) GetPrimaryExternalIDs() map[string]string {
// ...
}

func (id DBObjectIDs) GetSecondaryExternalIDs() map[string]string {
// ...
}

type DBObjectWithExternalIDs interface {
   GetExternalIDs() map[string]string
}

func (id DBObjectIDs) GetPrimaryIDPredicate[M DBObjectWithExternlaIDs](f func(item M) bool) func(item M) bool {
// must match all the primary IDs
}

func (id DBObjectIDs) GetSecondaryIDPredicate[M DBObjectWithExternlaIDs](f func(item M) bool) func(item M) bool {
// must match any of the secondary IDs
}

func (id DBObjectIDs) GetNoIDPredicate[M DBObjectWithExternlaIDs](f func(item M) bool) func(item M) bool {
// must verify that has no object IDs 
}
```

The external IDs woud look like
```
//		"k8s.ovn.org/object-type": "AddressSetType"
//		"k8s.ovn.org/controller": "controller-1"
//		"k8s.ovn.org/name": "namespace1_name1"
//              "k8s.ovn.org/family": "ipv4"
//		"k8s.ovn.org/id": "AddressSetType:controller-1:namespace1_name1:ipv4"
```

I am making this suggestion from the perspective that the current proposal seems overly complicated, but will add more comments with the specifics